# Eat Safe, Love

## Part 1: Database and Jupyter Notebook Set Up

Import the data provided in the `establishments.json` file from your Terminal. Name the database `uk_food` and the collection `establishments`.

Within this markdown cell, copy the line of text you used to import the data from your Terminal. This way, future analysts will be able to repeat your process.

e.g.: Import the dataset with `YOUR IMPORT TEXT HERE`

Import the dataset with `
mongoimport --type json -d uk_food -c establishments --drop --jsonArray "F:\github\nosql-challenge\Starter_Code\Resources\establishments.json"


In [1]:
# Import dependencies
from pymongo import MongoClient
from pprint import pprint

In [2]:
# Create an instance of MongoClient
mongo = MongoClient(port=27017)

In [3]:
# confirm that our new database was created
print(mongo.list_database_names())

['admin', 'autosaurus', 'classDB', 'class_db', 'config', 'epa', 'fruitsdb', 'local', 'met', 'petsitly_marketing', 'test', 'travel_db', 'uk_food']


In [4]:
# assign the uk_food database to a variable name
db = mongo['uk_food']

In [5]:
# review the collections in our new database
print(db.list_collection_names())

['establishments']


In [6]:
# review a document in the establishments collection
pprint(db.establishments.find_one())

{'AddressLine1': 'The Pines Garden',
 'AddressLine2': 'Beach Road',
 'AddressLine3': 'St Margarets Bay',
 'AddressLine4': 'Kent',
 'BusinessName': 'The Pines Calyx',
 'BusinessType': 'Other catering premises',
 'BusinessTypeID': 7841,
 'ChangesByServerID': 0,
 'Distance': 4587.362402580997,
 'FHRSID': 254250,
 'LocalAuthorityBusinessID': 'PI/000066174',
 'LocalAuthorityCode': '182',
 'LocalAuthorityEmailAddress': 'publicprotection@dover.gov.uk',
 'LocalAuthorityName': 'Dover',
 'LocalAuthorityWebSite': 'http://www.dover.gov.uk/',
 'NewRatingPending': False,
 'Phone': '',
 'PostCode': 'CT15 6DZ',
 'RatingDate': '2021-08-17T00:00:00',
 'RatingKey': 'fhrs_5_en-gb',
 'RatingValue': '5',
 'RightToReply': '',
 'SchemeType': 'FHRS',
 '_id': ObjectId('6741e3887a19a52e87e44220'),
 'geocode': {'latitude': '51.148133', 'longitude': '1.383298'},
 'links': [{'href': 'https://api.ratings.food.gov.uk/establishments/254250',
            'rel': 'self'}],
 'meta': {'dataSource': None,
          'extract

In [7]:
# assign the collection to a variable
establishments = db['establishments']

#### For memory, there are 39779 documents in the original DB

In [ ]:
# Display the number of documents : to check during the assignment all the changes

print(f'Number of documents is : {establishments.count_documents({})}')

Number of documents is : 39779


## Part 2: Update the Database

1. An exciting new halal restaurant just opened in Greenwich, but hasn't been rated yet. The magazine has asked you to include it in your analysis. Add the following restaurant "Penang Flavours" to the database.

In [ ]:
# Create a dictionary for the new restaurant data

new_restaurant = {
    "BusinessName":"Penang Flavours",
    "BusinessType":"Restaurant/Cafe/Canteen",
    "BusinessTypeID":"",
    "AddressLine1":"Penang Flavours",
    "AddressLine2":"146A Plumstead Rd",
    "AddressLine3":"London",
    "AddressLine4":"",
    "PostCode":"SE18 7DY",
    "Phone":"",
    "LocalAuthorityCode":"511",
    "LocalAuthorityName":"Greenwich",
    "LocalAuthorityWebSite":"http://www.royalgreenwich.gov.uk",
    "LocalAuthorityEmailAddress":"health@royalgreenwich.gov.uk",
    "scores":{
        "Hygiene":"",
        "Structural":"",
        "ConfidenceInManagement":""
    },
    "SchemeType":"FHRS",
    "geocode":{
        "longitude":"0.08384000",
        "latitude":"51.49014200"
    },
    "RightToReply":"",
    "Distance":4623.9723280747176,
    "NewRatingPending":True
}



In [ ]:
# Insert the new restaurant into the collection

establishments.insert_one(new_restaurant)

InsertOneResult(ObjectId('6741e3e927065827b86a1967'), acknowledged=True)

In [ ]:
# Check that the new restaurant was inserted : query on DB establishment > should return 1 result

query = {"BusinessName":"Penang Flavours"}
results = establishments.find(query)
for x in results : 
    pprint(x)

{'AddressLine1': 'Penang Flavours',
 'AddressLine2': '146A Plumstead Rd',
 'AddressLine3': 'London',
 'AddressLine4': '',
 'BusinessName': 'Penang Flavours',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': '',
 'Distance': 4623.972328074718,
 'LocalAuthorityCode': '511',
 'LocalAuthorityEmailAddress': 'health@royalgreenwich.gov.uk',
 'LocalAuthorityName': 'Greenwich',
 'LocalAuthorityWebSite': 'http://www.royalgreenwich.gov.uk',
 'NewRatingPending': True,
 'Phone': '',
 'PostCode': 'SE18 7DY',
 'RightToReply': '',
 'SchemeType': 'FHRS',
 '_id': ObjectId('6741e3e927065827b86a1967'),
 'geocode': {'latitude': '51.49014200', 'longitude': '0.08384000'},
 'scores': {'ConfidenceInManagement': '', 'Hygiene': '', 'Structural': ''}}


In [ ]:
# Display the number of documents : should be updated +1

print(f'Number of documents is : {establishments.count_documents({})}')

Number of documents is : 39780


2. Find the BusinessTypeID for "Restaurant/Cafe/Canteen" and return only the `BusinessTypeID` and `BusinessType` fields.

In [31]:
# Find the BusinessTypeID for "Restaurant/Cafe/Canteen" and return only the BusinessTypeID and BusinessType fields
query = {"BusinessType":"Restaurant/Cafe/Canteen"}
fields = {'BusinessTypeID': 1, 'BusinessType' :1, '_id' :0 }

# Capture the results to a variable

results = establishments.find(query, fields)

# Pretty print 
for i in range(1):

    # extract the business ID and store it in a variable

    result_id = results[i]['BusinessTypeID']
    
    # pprint the query result 

    pprint(results[i])
    print('')
    
    ## print the business ID
    
    print(f' The "BusinessTypeID" is : {result_id}')
    



{'BusinessType': 'Restaurant/Cafe/Canteen', 'BusinessTypeID': 1}

 The "BusinessTypeID" is : 1


3. Update the new restaurant with the `BusinessTypeID` you found.

In [14]:
# Update the new restaurant with the correct BusinessTypeID using variable result_id

db.establishments.update_one({"BusinessName":"Penang Flavours"}, \
                                { "$set": { "BusinessTypeID" : result_id}})

UpdateResult({'n': 1, 'nModified': 1, 'ok': 1.0, 'updatedExisting': True}, acknowledged=True)

In [ ]:
# Confirm that the new restaurant was updated : check key business ID of the query

query = {"BusinessName":"Penang Flavours"}
results = establishments.find(query)
for x in results : 
    pprint(x)
    print('')
    print(f'The new "BusinessTypeID" is {x['BusinessTypeID']}')

{'AddressLine1': 'Penang Flavours',
 'AddressLine2': '146A Plumstead Rd',
 'AddressLine3': 'London',
 'AddressLine4': '',
 'BusinessName': 'Penang Flavours',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': 1,
 'Distance': 4623.972328074718,
 'LocalAuthorityCode': '511',
 'LocalAuthorityEmailAddress': 'health@royalgreenwich.gov.uk',
 'LocalAuthorityName': 'Greenwich',
 'LocalAuthorityWebSite': 'http://www.royalgreenwich.gov.uk',
 'NewRatingPending': True,
 'Phone': '',
 'PostCode': 'SE18 7DY',
 'RightToReply': '',
 'SchemeType': 'FHRS',
 '_id': ObjectId('6741e3e927065827b86a1967'),
 'geocode': {'latitude': '51.49014200', 'longitude': '0.08384000'},
 'scores': {'ConfidenceInManagement': '', 'Hygiene': '', 'Structural': ''}}

The new "BusinessTypeID" is 1


In [ ]:
# # TO DELETE DUPLICATE DOCUMENTS (iF running the code several times)

# db.establishments.delete_one({"BusinessName":"Penang Flavours"})

# query = {"BusinessName":"Penang Flavours"}
# results = establishments.find(query)
# for x in results : 
#     print(x)

4. The magazine is not interested in any establishments in Dover, so check how many documents contain the Dover Local Authority. Then, remove any establishments within the Dover Local Authority from the database, and check the number of documents to ensure they were deleted.

In [ ]:
# Find how many documents have LocalAuthorityName as "Dover"

query_dover = {"LocalAuthorityName":"Dover"}
print(f' Number of documents with LocalAuthorityName as "Dover" is : {establishments.count_documents(query_dover)}')



 Number of documents with LocalAuthorityName as "Dover" is : 994


##### for memory number of documents with Dover needs to be : 994

In [ ]:
# Capture the results to a variable

Dover_restaurants = establishments.find(query_dover)


# Pretty print the first two results

for i in range(2):
    pprint(Dover_restaurants[i])

{'AddressLine1': 'The Pines Garden',
 'AddressLine2': 'Beach Road',
 'AddressLine3': 'St Margarets Bay',
 'AddressLine4': 'Kent',
 'BusinessName': 'The Pines Calyx',
 'BusinessType': 'Other catering premises',
 'BusinessTypeID': 7841,
 'ChangesByServerID': 0,
 'Distance': 4587.362402580997,
 'FHRSID': 254250,
 'LocalAuthorityBusinessID': 'PI/000066174',
 'LocalAuthorityCode': '182',
 'LocalAuthorityEmailAddress': 'publicprotection@dover.gov.uk',
 'LocalAuthorityName': 'Dover',
 'LocalAuthorityWebSite': 'http://www.dover.gov.uk/',
 'NewRatingPending': False,
 'Phone': '',
 'PostCode': 'CT15 6DZ',
 'RatingDate': '2021-08-17T00:00:00',
 'RatingKey': 'fhrs_5_en-gb',
 'RatingValue': '5',
 'RightToReply': '',
 'SchemeType': 'FHRS',
 '_id': ObjectId('6741e3887a19a52e87e44220'),
 'geocode': {'latitude': '51.148133', 'longitude': '1.383298'},
 'links': [{'href': 'https://api.ratings.food.gov.uk/establishments/254250',
            'rel': 'self'}],
 'meta': {'dataSource': None,
          'extract

In [19]:
# Delete all documents where LocalAuthorityName is "Dover"

db.establishments.delete_many(query_dover)

DeleteResult({'n': 994, 'ok': 1.0}, acknowledged=True)

##### number of documents without Dover should be : 38786

In [20]:
# Display the number of remaining documents

print(f'Number of remaining documents is : {establishments.count_documents({})}')

Number of remaining documents is : 38786


In [ ]:
# Check if any remaining documents include Dover : should be 0

query = {"LocalAuthorityName":"Dover"}
print(f' Number of remaining documents with "LocalAuthorityName" as "Dover" is : {establishments.count_documents(query)}')

 Number of remaining documents with "LocalAuthorityName" as "Dover" is : 0


In [22]:
# Check that other documents remain with 'find_one'
results = establishments.find_one()

pprint(results)

{'AddressLine1': 'East Cliff Pavilion',
 'AddressLine2': 'Wear Bay Road',
 'AddressLine3': 'Folkestone',
 'AddressLine4': 'Kent',
 'BusinessName': 'The Pavilion',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': 1,
 'ChangesByServerID': 0,
 'Distance': 4591.765489457773,
 'FHRSID': 1043695,
 'LocalAuthorityBusinessID': 'PI/000073616',
 'LocalAuthorityCode': '188',
 'LocalAuthorityEmailAddress': 'foodteam@folkestone-hythe.gov.uk',
 'LocalAuthorityName': 'Folkestone and Hythe',
 'LocalAuthorityWebSite': 'http://www.folkestone-hythe.gov.uk',
 'NewRatingPending': False,
 'Phone': '',
 'PostCode': 'CT19 6BL',
 'RatingDate': '2018-04-04T00:00:00',
 'RatingKey': 'fhrs_5_en-gb',
 'RatingValue': '5',
 'RightToReply': '',
 'SchemeType': 'FHRS',
 '_id': ObjectId('6741e3887a19a52e87e44507'),
 'geocode': {'latitude': '51.083812', 'longitude': '1.195625'},
 'links': [{'href': 'https://api.ratings.food.gov.uk/establishments/1043695',
            'rel': 'self'}],
 'meta': {'dataSource': 

5. Some of the number values are stored as strings, when they should be stored as numbers.

Use `update_many` to convert `latitude` and `longitude` to decimal numbers.

In [23]:
# Change the data type from String to Decimal for longitude and latitude

establishments.update_many({}, [ {'$set': { "geocode.longitude" : {'$toDouble': "$geocode.longitude"},
                                            "geocode.latitude" : {'$toDouble': "$geocode.latitude"}
                                              }
                                     } ]
                              )

UpdateResult({'n': 38786, 'nModified': 38786, 'ok': 1.0, 'updatedExisting': True}, acknowledged=True)

Use `update_many` to convert `RatingValue` to integer numbers.

In [24]:
# Set non 1-5 Rating Values to Null  : I CHANGED Null to 0 , there should be a warning to inform the reader that 0 = not rated

non_ratings = ["AwaitingInspection", "Awaiting Inspection", "AwaitingPublication", "Pass", "Exempt"]
establishments.update_many({"RatingValue": {"$in": non_ratings}}, [ {'$set':{ "RatingValue" : "0"}} ])

UpdateResult({'n': 4091, 'nModified': 4091, 'ok': 1.0, 'updatedExisting': True}, acknowledged=True)

In [25]:
establishments.update_many({}, [ {'$set': { "RatingValue" : {'$toInt': "$RatingValue"}
                                      } } ] )

UpdateResult({'n': 38786, 'nModified': 38786, 'ok': 1.0, 'updatedExisting': True}, acknowledged=True)

In [26]:
# test the conversion of rating 

query_rating = {"RatingValue":{'$gt' : 1 }}
print(f' Number of documents with RatingValue is : {establishments.count_documents(query_rating)}')


 Number of documents with RatingValue is : 34059


In [27]:
# Check that the coordinates and rating value are now numbers
results = establishments.find_one()


print(f' Latitude & longitude data type is now : {results['geocode']}')

print(f' Rating value is now : {results['RatingValue']}')

 Latitude & longitude data type is now : {'longitude': 1.195625, 'latitude': 51.083812}
 Rating value is now : 5


In [ ]:
# Check that the coordinates and rating value are now numbers
pprint(results)

{'AddressLine1': 'East Cliff Pavilion',
 'AddressLine2': 'Wear Bay Road',
 'AddressLine3': 'Folkestone',
 'AddressLine4': 'Kent',
 'BusinessName': 'The Pavilion',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': 1,
 'ChangesByServerID': 0,
 'Distance': 4591.765489457773,
 'FHRSID': 1043695,
 'LocalAuthorityBusinessID': 'PI/000073616',
 'LocalAuthorityCode': '188',
 'LocalAuthorityEmailAddress': 'foodteam@folkestone-hythe.gov.uk',
 'LocalAuthorityName': 'Folkestone and Hythe',
 'LocalAuthorityWebSite': 'http://www.folkestone-hythe.gov.uk',
 'NewRatingPending': False,
 'Phone': '',
 'PostCode': 'CT19 6BL',
 'RatingDate': '2018-04-04T00:00:00',
 'RatingKey': 'fhrs_5_en-gb',
 'RatingValue': 5,
 'RightToReply': '',
 'SchemeType': 'FHRS',
 '_id': ObjectId('6741e3887a19a52e87e44507'),
 'geocode': {'latitude': 51.083812, 'longitude': 1.195625},
 'links': [{'href': 'https://api.ratings.food.gov.uk/establishments/1043695',
            'rel': 'self'}],
 'meta': {'dataSource': None,
